drivers/gpio/gpio_sam4l.c

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright (c) 2018 Justin Watson
+ * Copyright (c) 2020 Gerson Fernando Budke <nandojve@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT atmel_sam4l_gpio
+
+#include <errno.h>
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <soc.h>
+#include <drivers/gpio.h>
+
+#include "gpio_utils.h"
+
+typedef void (*config_func_t)(const struct device *dev);
+
+struct gpio_sam_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	Gpio *regs;
+	config_func_t config_func;
+	uint32_t periph_id;
+};
+
+struct gpio_sam_runtime {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	sys_slist_t cb;
+};
+
+#define DEV_CFG(dev) \
+	((const struct gpio_sam_config * const)(dev)->config)
+#define DEV_DATA(dev) \
+	((struct gpio_sam_runtime * const)(dev)->data)
+
+#define GPIO_SAM_ALL_PINS    0xFFFFFFFF
+
+static int gpio_sam_port_configure(const struct device *dev,
+				   uint32_t mask,
+				   gpio_flags_t flags)
+{
+	const struct gpio_sam_config * const cfg = DEV_CFG(dev);
+	Gpio * const gpio = cfg->regs;
+
+	/* No hardware support */
+	if ((flags & GPIO_SINGLE_ENDED) ||
+	    ((flags & GPIO_PULL_UP) && (flags & GPIO_PULL_DOWN))) {
```

Comment:
There is an [assert in gpio.h](https://github.com/zephyrproject-rtos/zephyr/blob/dd5eafcbc70718cc38013c67b46786be87aef3ef/include/drivers/gpio.h#L553) guarding this condition. We could remove it from here.